rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+# SlateDB Transaction
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [SlateDB Transaction](#slatedb-transaction)
+  - [Background](#background)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Constraints](#constraints)
+  - [References](#references)
+  - [Proposal](#proposal)
+    - [API](#api)
+    - [Isolation Levels](#isolation-levels)
+    - [Prerequisites](#prerequisites)
+      - [Snapshot](#snapshot)
+      - [Write Batch](#write-batch)
+  - [Conflict Checking: Snapshot Isolation](#conflict-checking-snapshot-isolation)
+  - [Conflict Checking: Serializable Snapshot Isolation](#conflict-checking-serializable-snapshot-isolation)
+  - [Roadmap](#roadmap)
+  - [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Reviewing
+
+Authors:
+
+* [Li Yazhou](https://github.com/flaneur2020)
+
+## Background
+
+Transaction allows multiple operations to be executed in a single atomic operation, and control the isolation level of the operations. This is a planned feature in SlateDB, and is essential for many critical use cases, like:
+
+- Using SlateDB as a metadata store or coordinator for a distributed system, such as the backend of [Kine](https://github.com/k3s-io/kine) or various metadata stores in OLAP systems.
+- Using SlateDB as a underlying KV layer in an OLTP system, such as a [TiKV](https://tikv.org/) replacement or other SQL engines.
+
+This RFC proposes the goals & design draft of the transaction feature in SlateDB.
+
+There're still some ongoing RFCs on the prerequisites of the Transaction feature, such as the Snapshot API and WriteBatch API. We'll refer to these RFCs or PRs in this RFC when they're ready.
+
+Is this RFC considered as too early? Yes, the implementation of the Transaction feature can not be started until the prerequisites got ready. However, this RFC may help us to clarify the requirements of the Snapshot API and WriteBatch API in the sense of the Transaction feature, and organize the roadmap on these developments.
```

Comment:
Nit: This RFC does very concretely proposes how transactions will be implemented, it also clearly brings out the requirements and shared the roadmap. Until I read the full RFC, this sentence made me think that this is more of an exploration. Looks like it is a concrete proposal, can this be reworded?

Reply:
thank you for the suggestion, it worths a reword. as you suggested, the structure of this doc also worths a refactor. let me figure out a way on this.